fault_proving(global_roots): scaffold global merkle root storage crate

netrome opened the pull request:
Closes #2552

This PR adds a new crate with column definitions for the global state root service, which we intend to deploy as a standalone binary.

Moreover, this crate also contains a top-level function to populate these columns with data given a new block.

xgreenx commented on crates/storage/src/column.rs:
```diff
@@ -71,6 +71,12 @@ pub enum Column {
     // TODO: Remove this column and use `Metadata` column instead.
     /// Table for genesis state import progress tracking.
     GenesisMetadata = 21,
+    /// The Merkle Nodes for the Coins Table
+    #[cfg(feature = "global-state-root")]
+    CoinsMerkleData = 22,
+    /// The Merkle Root for the Coins Table
+    #[cfg(feature = "global-state-root")]
+    CoinsMerkleMetadata = 23,
```
For Global Merkle root we can create a new `Column` type. We don't need to re-use onchain column type

netrome replied:
Done in https://github.com/FuelLabs/fuel-core/pull/2553/commits/13521ac08c857a093836a9bbf02f43986099a89c (by you ;) )

xgreenx commented on crates/storage/src/structured_storage/coins.rs:
```diff
@@ -11,12 +10,115 @@ use crate::{
-impl TableWithBlueprint for Coins {
-    type Blueprint = Plain<Primitive<34>, Postcard>;
-    type Column = Column;
+#[cfg(not(feature = "global-state-root"))]
+mod plain {
```
We don't need to integrate it into the `fuel-core-storage`. It should be a new stance alone crate that can accept the block as an input and will create Merkle roots over data from there=D

netrome replied:
Done in https://github.com/FuelLabs/fuel-core/pull/2553/commits/13521ac08c857a093836a9bbf02f43986099a89c (by you ;) )

netrome commented:
Redefined the PR and issue after @xgreenx's proposal. I will do some further cleanups and create follow-up issues for the next steps, before marking this PR as ready for review.

xgreenx commented on crates/fraud_proofs/global_merkle_root/storage/Cargo.toml:
```diff
@@ -0,0 +1,35 @@
+[package]
+authors = { workspace = true }
+categories = ["cryptography::cryptocurrencies"]
+description = "Storage types and primitives used by Fuel core protocol."
```
It would be nice to update crate related stuff, because I just copied `fuel-core-storage`=D

netrome replied:
Good point, cleaned it up in https://github.com/FuelLabs/fuel-core/pull/2553/commits/02ee8224cba69791fbfab1c58aeaae65c575e41d

xgreenx commented on crates/fraud_proofs/global_merkle_root/storage/src/update.rs:
```diff
@@ -0,0 +1,311 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+#![deny(clippy::arithmetic_side_effects)]
+#![deny(clippy::cast_possible_truncation)]
+// #![deny(unused_crate_dependencies)]
+// #![deny(missing_docs)]
+// #![deny(warnings)]
+
+#[cfg(feature = "alloc")]
+extern crate alloc;
```
I htink we don't need these=)

netrome replied:
Yeah, no only in the root `lib.rs`. Removed them in https://github.com/FuelLabs/fuel-core/pull/2553/commits/9e7658e6314b2b8d696e2c1edd3c0653a988dbf2

xgreenx reviewed (commented):
It would be nice to verify that new crate us not std compatible by updating our CI and `./ci_check.sh` script=)

netrome commented:
> It would be nice to verify that new crate us not std compatible by updating our CI and `./ci_check.sh` script=)

Good point, fixed the no-std build and updated the CI checks in https://github.com/FuelLabs/fuel-core/pull/2553/commits/e0bd85c5ffbe3e159d59d201d4b67509d748c17e and https://github.com/FuelLabs/fuel-core/pull/2553/commits/1c1404daa18555d2259ebf8ad8a2147730c9f183

rymnc commented on ci_checks.sh:
```diff
@@ -27,6 +27,7 @@ cargo check -p fuel-core-storage --target wasm32-unknown-unknown --no-default-fe
 cargo check -p fuel-core-client --target wasm32-unknown-unknown --no-default-features &&
 cargo check -p fuel-core-chain-config --target wasm32-unknown-unknown --no-default-features &&
 cargo check -p fuel-core-executor --target wasm32-unknown-unknown --no-default-features --features alloc &&
+cargo check -p fuel-core-global-merkle-root-storage --target wasm32-unknown-unknown --no-default-features &&
```
@netrome i think green meant to add it to the `.github/workflows/ci.yml` file too :)

netrome replied:
Right, updated in https://github.com/FuelLabs/fuel-core/pull/2553/commits/83f9cc92e1f8f0fa53172d51f7bc90cdbadfe2c1 thanks for pointing it out :pray:

acerone85 commented on crates/fraud_proofs/global_merkle_root/storage/src/merkle/smt.rs:
```diff
@@ -0,0 +1,60 @@
+
+impl<Table> TableWithBlueprint for MerkleData<Table>
+where
+    Table: MerklizedTableColumn,
```
nit, but I believe the correct spelling is `Merkleized`
```suggestion
    Table: MerkleizedTableColumn,
```

netrome replied:
Sure, I'll update.

netrome replied:
Done in https://github.com/FuelLabs/fuel-core/pull/2553/commits/6b0d8196f1fa215fba6007e83f3edb5622f575bc

acerone85 commented on crates/fraud_proofs/global_merkle_root/storage/src/column.rs:
```diff
@@ -0,0 +1,114 @@
+
+impl Column {
+    /// The total count of variants in the enum.
+    pub const COUNT: usize = TableColumn::COUNT + TableColumn::COUNT;
```
shouldn't this be `TableColumn::COUNT + TableColumn::COUNT + TableColumn::COUNT`?

netrome replied:
Right, it should. I'd just formulate it as `TableColumn::COUNT * 3`

netrome replied:
I just cherry-picked your fix instead. Thank you!

acerone85 commented on crates/fraud_proofs/global_merkle_root/storage/src/column.rs:
```diff
@@ -0,0 +1,114 @@
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum Column {
+    TableColumn(TableColumn),
+    MerkleMetadataColumn(TableColumn),
```
Each table has its own KeyConverter that converts a key to its column. So in theory we could do away with only one `MerkleMetadataColumn`?

netrome replied:
I think so, as long as we don't risk any collisions. I see you've already proposed a change in https://github.com/FuelLabs/fuel-core/pull/2611. It looks good to me, I'll cherry-pick it into this PR.

netrome replied:
Done here: https://github.com/FuelLabs/fuel-core/pull/2553/commits/bc1209305e66ce5584e9af2a8b7cc9abbb546f59

acerone85 commented on crates/fraud_proofs/global_merkle_root/storage/src/merkle/smt.rs:
```diff
@@ -0,0 +1,60 @@
+    type Blueprint = Plain<Primitive<4>, Postcard>;
+    type Column = Column;
+
+    fn column() -> Column {
+        Column::MerkleDataColumn(Table::table_column())
+    }
```
Should this be the `MerkleMetadataColumn`, or am I missing something?
```suggestion
    fn column() -> Column {
        Column::MerkleMetadataColumn(Table::table_column())
    }
```

rafal-ch commented on crates/fraud_proofs/global_merkle_root/storage/src/column.rs:
```diff
@@ -0,0 +1,114 @@
+    /// Only can be proved with list of events processed during the block
+    /// and compared with the `event_inbox_root` in the block header.
+    Messages = 3,
+    /// We need to prove that the transaction doesn't included into the table.
```
A nit:

```suggestion
    /// We need to prove that the transaction isn't included in the table.
```

rafal-ch commented on crates/fraud_proofs/global_merkle_root/storage/src/column.rs:
```diff
@@ -0,0 +1,114 @@
+
+impl TableColumn {
+    /// The total count of variants in the enum.
+    pub const COUNT: usize = <Self as strum::EnumCount>::COUNT;
```
It looks like this is needed only within the module.

```suggestion
    const COUNT: usize = <Self as strum::EnumCount>::COUNT;
```

rafal-ch commented on crates/fraud_proofs/global_merkle_root/storage/src/column.rs:
```diff
@@ -0,0 +1,114 @@
+                let str: &str = column.into();
+                str.to_string()
+            }
+            Column::MerkleMetadataColumn(column) => {
```
A nit, for the sake of consistency :)

```suggestion
            Self::MerkleMetadataColumn(column) => {
```

rafal-ch commented on crates/fraud_proofs/global_merkle_root/storage/src/lib.rs:
```diff
@@ -0,0 +1,27 @@
+pub type StateTransitionBytecodeVersions =
+    Merklized<fuel_core_storage::tables::StateTransitionBytecodeVersions>;
+pub type UploadedBytecodes = Merklized<fuel_core_storage::tables::UploadedBytecodes>;
+pub type Blobs = Merklized<fuel_core_storage::tables::BlobData>;
```
Why not `pub type BlobData...`?
Not a big deal, but this is the only type that stands out :-)

netrome replied:
Yeah this follows the naming in the original storage crate, where the column is `Blobs` although the table is called `BlobData`. It's a bit weird, but I think `Blobs` sound better here.

rafal-ch commented on crates/fraud_proofs/global_merkle_root/storage/src/update.rs:
```diff
@@ -0,0 +1,305 @@
+            }
+            .into();
+
+            self.storage.storage::<Coins>().insert(&utxo_id, &coin)?;
```
Not sure, so just leaving the comment here.

Should we consider using `replace` here and check if there's been a coin in the storage already?

netrome replied:
Hmm, I think this is fine but it wouldn't hurt to be extra conservative. There should never be a coin that we overwrite, so sure let's do it.

netrome replied:
Done in https://github.com/FuelLabs/fuel-core/pull/2553/commits/6aa1c2f2b436b056c6a78ee88796c933fe6972ad

acerone85 commented:
one of my commits broke the CI XD
but that aside, it looks good to me. 
Happy to approve and reapprove once the CI passes

netrome commented:
> one of my commits broke the CI XD but that aside, it looks good to me. Happy to approve and reapprove once the CI passes

Oh I see, I fixed it now. Would love to get some re-approvals.
